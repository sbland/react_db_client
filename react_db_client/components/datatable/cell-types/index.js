export { DataTableCellButton } from './DataTableCellButton';
export { DataTableCellEntity } from './DataTableCellEntity';
export { DataTableCellLink } from './DataTableCellLink';
export { DataTableCellNumber } from './DataTableCellNumber';
export { DataTableCellPopup } from './DataTableCellPopup';
export { DataTableCellReadOnly } from './DataTableCellReadOnly';
export { DataTableCellSelect } from './DataTableCellSelect';
export { DataTableCellText } from './DataTableCellText';
export { DataTableCellToggle } from './DataTableCellToggle';

export { defaultComponentMap } from './default-component-map';
