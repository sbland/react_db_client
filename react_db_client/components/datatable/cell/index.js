export { Cell } from './cell';
export { DataTableCellHoverWrap } from './cell-wrappers';
