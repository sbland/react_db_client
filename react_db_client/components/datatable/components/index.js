// export { DataTableCellHoverWrap } from './DataTableCell';
export { DataTableTotals } from './DataTableTotals';
export { DataTableHeadings } from './DataTableHeadings';
