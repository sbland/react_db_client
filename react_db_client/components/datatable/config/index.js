export {
  dataTableDefaultConfig,
  DataTableContext,
  DataTableConfigConnector,
} from './DataTableConfig';
