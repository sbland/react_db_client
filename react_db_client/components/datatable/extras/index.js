export { RowErrors } from './errorTypes';
export { headingDataShape, tableDataShape } from './inputDataShapes';
export * from './demoData';