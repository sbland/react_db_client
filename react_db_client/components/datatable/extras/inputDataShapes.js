import PropTypes from 'prop-types';

export const headingDataShape = {
  uid: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  action: PropTypes.func,
  to: PropTypes.string,
};
/** Table Data Prop Type Shape
 * ... plus map against headings
 */
export const tableDataShape = {
  uid: PropTypes.string.isRequired,
  name: PropTypes.string,
  // ... plus map against headings
};
