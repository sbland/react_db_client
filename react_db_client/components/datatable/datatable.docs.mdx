---
description: 'A datatable spreadsheet component.'
labels: ['data', 'ui']
---

import { Datatable } from './datatable';

## React Component for displaying and editing table data

A simple spreadsheet component

