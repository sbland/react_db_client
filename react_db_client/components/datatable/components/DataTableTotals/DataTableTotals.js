import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { headingDataShape } from '@samnbuk/react_db_client.components.datatable.extras';
import { DataTableContext } from '@samnbuk/react_db_client.components.datatable.logic';
import { formatValue } from '@samnbuk/react_db_client.helpers.data-processing';
import {
  DataTableCellHoverWrap,
  DataTableDataCell,
} from '@samnbuk/react_db_client.components.datatable.cell-types';

export const DataTableTotals = ({
  className,
  headingsDataList,
  columnWidths,
  totals,
  tableWidth,
  componentMap,
}) => {
  const { hasBtnsColumn } = useContext(DataTableContext);

  const insideWrapStyleOverride = {
    width: `${tableWidth}px`,
  };

  const mapTotals = headingsDataList.map((headingData, i) => {
    const cellData = headingData.showTotals ? (
      formatValue(Number(totals[headingData.uid]), headingData.step || 0.01)
    ) : (
      <div />
    );
    return (
      <DataTableCellHoverWrap
        key={headingData.uid}
        columnWidth={columnWidths[hasBtnsColumn ? i + 1 : i]}
        className="dataTableTotals"
      >
        {headingData.showTotals ? (
          <DataTableDataCell
            rowId="_totals"
            rowIndex={-1}
            rowData={{}}
            classNames={className}
            columnId={headingData.uid}
            cellData={cellData}
            columnData={headingData}
            // cellDataBtnRef={cellDataBtnRef}
            componentMap={componentMap}
            updateData={() => {}} // included for future
            acceptValue={() => {}} // included for future
            resetValue={() => {}} // included for future
            focused={false} // included for future
            editMode={false} // included for future
            isDisabled={false}
          />
        ) : (
          <div />
        )}
      </DataTableCellHoverWrap>
    );
  });

  if (hasBtnsColumn) {
    mapTotals.unshift(
      <DataTableCellHoverWrap
        className="dataTableTotals"
        key="Settings"
        columnWidth={columnWidths[0]}
      >
        <></>
      </DataTableCellHoverWrap>
    );
  }

  return (
    <div className="dataTable_totalsOutsideWrap">
      <div className="dataTable_totalsInsideWrap" style={insideWrapStyleOverride}>
        {mapTotals}
      </div>
    </div>
  );
};

DataTableTotals.propTypes = {
  headingsDataList: PropTypes.arrayOf(PropTypes.shape(headingDataShape)).isRequired,
  totals: PropTypes.objectOf(PropTypes.number).isRequired,
  columnWidths: PropTypes.arrayOf(PropTypes.number).isRequired,
  tableWidth: PropTypes.number.isRequired,
  componentMap: PropTypes.objectOf(PropTypes.elementType).isRequired,
  className: PropTypes.string,
};

DataTableTotals.defaultProps = {
  className: '',
};

export default DataTableTotals;
