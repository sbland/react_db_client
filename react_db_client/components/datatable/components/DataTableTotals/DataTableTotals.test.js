import React from 'react';
import { mount, shallow } from 'enzyme';

import { demoHeadingsData } from '@samnbuk/react_db_client.components.datatable.extras';
import { DataTableTotals } from './DataTableTotals';
import {
  DataTableCellHoverWrap,
  DataTableDataCell,
} from '@samnbuk/react_db_client.components.datatable.cell';

const defaultProps = {
  headingsDataList: demoHeadingsData,
  columnWidths: demoHeadingsData.map(() => 10),
  totals: demoHeadingsData.map(() => 10),
  tableWidth: demoHeadingsData.reduce((acc, v) => acc + v),
  customFieldComponents: {},
};

describe('DataTableTotals', () => {
  beforeEach(() => {});
  test('Renders', () => {
    shallow(<DataTableTotals {...defaultProps} />);
  });
  describe('shallow renders', () => {
    test('Matches Snapshot', () => {
      const component = shallow(<DataTableTotals {...defaultProps} />);
      const tree = component.debug();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('Unit tests', () => {
    let component;
    beforeEach(() => {
      component = mount(<DataTableTotals {...defaultProps} />);
    });

    describe('Showing Totals', () => {
      test('should show totals for columns with showTotals set to true', () => {
        const totalsRow = component.find(DataTableCellHoverWrap);
        const totalsCells = totalsRow.find(DataTableDataCell);
        expect(totalsCells).toHaveLength(demoHeadingsData.filter((h) => h.showTotals).length); // Not sure why there are double here
        expect(totalsCells.debug()).toMatchSnapshot();
      });
    });
  });
});
