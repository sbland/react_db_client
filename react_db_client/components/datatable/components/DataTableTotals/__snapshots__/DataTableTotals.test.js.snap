// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTableTotals Unit tests Showing Totals should show totals for columns with showTotals set to true 1`] = `
"<DataTableDataCell rowId=\\"_totals\\" rowIndex={-1} rowData={{...}} classNames=\\"\\" columnId=\\"count\\" cellData={0} columnData={{...}} customFieldComponents={{...}} updateData={[Function: updateData]} acceptValue={[Function: acceptValue]} resetValue={[Function: resetValue]} focused={false} editMode={false} isDisabled={false}>
  <div style={{...}}>
    <CellRightClickWrapper readOnly={false} clearCell={[Function: clearCell]} setAsDefault={[Function: setAsDefault]}>
      <RightClickWrapper items={{...}}>
        <div style={{...}} className=\\"popupMenu_righClickWrap false\\" onContextMenu={[Function: handleContextMenu]}>
          <DataTableCellNumber rowId=\\"_totals\\" rowIndex={-1} rowData={{...}} classNames=\\"\\" columnId=\\"count\\" cellData={0} columnData={{...}} customFieldComponents={{...}} updateData={[Function: updateData]} acceptValue={[Function: acceptValue]} resetValue={[Function: resetValue]} focused={false} editMode={false} isDisabled={false}>
            <div className=\\"dataTableCellData dataTableCellData-number\\">
              <div className=\\"dataTableCellData_text\\">
                0
              </div>
              <input style={{...}} className=\\"cellInput-number\\" autoFocus={true} onFocus={[Function: onFocus]} type=\\"number\\" max={10} min={2} onChange={[Function: handleInputChange]} value=\\"\\" step={1} onBlur={[Function: onBlur]} onKeyDown={[Function: onKeyPress]} />
            </div>
          </DataTableCellNumber>
        </div>
        <PopupMenu isOpenOverride={false} onCloseCallback={[Function: onCloseCallback]} items={{...}} position={{...}} />
      </RightClickWrapper>
    </CellRightClickWrapper>
  </div>
</DataTableDataCell>


<DataTableDataCell rowId=\\"_totals\\" rowIndex={-1} rowData={{...}} classNames=\\"\\" columnId=\\"def\\" cellData={0} columnData={{...}} customFieldComponents={{...}} updateData={[Function: updateData]} acceptValue={[Function: acceptValue]} resetValue={[Function: resetValue]} focused={false} editMode={false} isDisabled={false}>
  <div style={{...}}>
    <CellRightClickWrapper readOnly={false} clearCell={[Function: clearCell]} setAsDefault={[Function: setAsDefault]}>
      <RightClickWrapper items={{...}}>
        <div style={{...}} className=\\"popupMenu_righClickWrap false\\" onContextMenu={[Function: handleContextMenu]}>
          <DataTableCellNumber rowId=\\"_totals\\" rowIndex={-1} rowData={{...}} classNames=\\"\\" columnId=\\"def\\" cellData={0} columnData={{...}} customFieldComponents={{...}} updateData={[Function: updateData]} acceptValue={[Function: acceptValue]} resetValue={[Function: resetValue]} focused={false} editMode={false} isDisabled={false}>
            <div className=\\"dataTableCellData dataTableCellData-number\\">
              <div className=\\"dataTableCellData_text\\">
                0
              </div>
              <input style={{...}} className=\\"cellInput-number\\" autoFocus={true} onFocus={[Function: onFocus]} type=\\"number\\" max={10} min={1} onChange={[Function: handleInputChange]} value=\\"\\" step={0.01} onBlur={[Function: onBlur]} onKeyDown={[Function: onKeyPress]} />
            </div>
          </DataTableCellNumber>
        </div>
        <PopupMenu isOpenOverride={false} onCloseCallback={[Function: onCloseCallback]} items={{...}} position={{...}} />
      </RightClickWrapper>
    </CellRightClickWrapper>
  </div>
</DataTableDataCell>


<DataTableDataCell rowId=\\"_totals\\" rowIndex={-1} rowData={{...}} classNames=\\"\\" columnId=\\"expression\\" cellData={0} columnData={{...}} customFieldComponents={{...}} updateData={[Function: updateData]} acceptValue={[Function: acceptValue]} resetValue={[Function: resetValue]} focused={false} editMode={false} isDisabled={false}>
  <div style={{...}}>
    <CellRightClickWrapper readOnly={false} clearCell={[Function: clearCell]} setAsDefault={[Function: setAsDefault]}>
      <RightClickWrapper items={{...}}>
        <div style={{...}} className=\\"popupMenu_righClickWrap false\\" onContextMenu={[Function: handleContextMenu]}>
          <DataTableCellNumber rowId=\\"_totals\\" rowIndex={-1} rowData={{...}} classNames=\\"\\" columnId=\\"expression\\" cellData={0} columnData={{...}} customFieldComponents={{...}} updateData={[Function: updateData]} acceptValue={[Function: acceptValue]} resetValue={[Function: resetValue]} focused={false} editMode={false} isDisabled={false}>
            <div className=\\"dataTableCellData dataTableCellData-number\\">
              <div className=\\"dataTableCellData_text\\">
                0
              </div>
              <input style={{...}} className=\\"cellInput-number\\" autoFocus={true} onFocus={[Function: onFocus]} type=\\"number\\" max={[undefined]} min={[undefined]} onChange={[Function: handleInputChange]} value=\\"\\" step={0.01} onBlur={[Function: onBlur]} onKeyDown={[Function: onKeyPress]} />
            </div>
          </DataTableCellNumber>
        </div>
        <PopupMenu isOpenOverride={false} onCloseCallback={[Function: onCloseCallback]} items={{...}} position={{...}} />
      </RightClickWrapper>
    </CellRightClickWrapper>
  </div>
</DataTableDataCell>


<DataTableDataCell rowId=\\"_totals\\" rowIndex={-1} rowData={{...}} classNames=\\"\\" columnId=\\"customfield\\" cellData={0} columnData={{...}} customFieldComponents={{...}} updateData={[Function: updateData]} acceptValue={[Function: acceptValue]} resetValue={[Function: resetValue]} focused={false} editMode={false} isDisabled={false}>
  <div style={{...}}>
    <CellRightClickWrapper readOnly={false} clearCell={[Function: clearCell]} setAsDefault={[Function: setAsDefault]}>
      <RightClickWrapper items={{...}}>
        <div onContextMenu={[Function: handleContextMenu]}>
          Invalid Cell Type 
          custom
        </div>
        <PopupMenu isOpenOverride={false} onCloseCallback={[Function: onCloseCallback]} items={{...}} position={{...}} />
      </RightClickWrapper>
    </CellRightClickWrapper>
  </div>
</DataTableDataCell>


<DataTableDataCell rowId=\\"_totals\\" rowIndex={-1} rowData={{...}} classNames=\\"\\" columnId=\\"customfieldeval\\" cellData={0} columnData={{...}} customFieldComponents={{...}} updateData={[Function: updateData]} acceptValue={[Function: acceptValue]} resetValue={[Function: resetValue]} focused={false} editMode={false} isDisabled={false}>
  <div style={{...}}>
    <CellRightClickWrapper readOnly={false} clearCell={[Function: clearCell]} setAsDefault={[Function: setAsDefault]}>
      <RightClickWrapper items={{...}}>
        <div onContextMenu={[Function: handleContextMenu]}>
          Invalid Cell Type 
          customeval
        </div>
        <PopupMenu isOpenOverride={false} onCloseCallback={[Function: onCloseCallback]} items={{...}} position={{...}} />
      </RightClickWrapper>
    </CellRightClickWrapper>
  </div>
</DataTableDataCell>"
`;

exports[`DataTableTotals shallow renders Matches Snapshot 1`] = `
"<div className=\\"dataTable_totalsOutsideWrap\\">
  <div className=\\"dataTable_totalsInsideWrap\\" style={{...}}>
    <DataTableCellHoverWrap className=\\"dataTableTotals\\" columnWidth={10} style={{...}} handleHover={[Function: handleHover]} disabled={false} />
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <DataTableDataCell rowId=\\"_totals\\" rowIndex={-1} rowData={{...}} classNames=\\"\\" columnId=\\"count\\" cellData={0} columnData={{...}} customFieldComponents={{...}} updateData={[Function: updateData]} acceptValue={[Function: acceptValue]} resetValue={[Function: resetValue]} focused={false} editMode={false} isDisabled={false} />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <DataTableDataCell rowId=\\"_totals\\" rowIndex={-1} rowData={{...}} classNames=\\"\\" columnId=\\"def\\" cellData={0} columnData={{...}} customFieldComponents={{...}} updateData={[Function: updateData]} acceptValue={[Function: acceptValue]} resetValue={[Function: resetValue]} focused={false} editMode={false} isDisabled={false} />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <DataTableDataCell rowId=\\"_totals\\" rowIndex={-1} rowData={{...}} classNames=\\"\\" columnId=\\"expression\\" cellData={0} columnData={{...}} customFieldComponents={{...}} updateData={[Function: updateData]} acceptValue={[Function: acceptValue]} resetValue={[Function: resetValue]} focused={false} editMode={false} isDisabled={false} />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <div />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={10} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <DataTableDataCell rowId=\\"_totals\\" rowIndex={-1} rowData={{...}} classNames=\\"\\" columnId=\\"customfield\\" cellData={0} columnData={{...}} customFieldComponents={{...}} updateData={[Function: updateData]} acceptValue={[Function: acceptValue]} resetValue={[Function: resetValue]} focused={false} editMode={false} isDisabled={false} />
    </DataTableCellHoverWrap>
    <DataTableCellHoverWrap columnWidth={100} className=\\"dataTableTotals\\" style={{...}} handleHover={[Function: handleHover]} disabled={false}>
      <DataTableDataCell rowId=\\"_totals\\" rowIndex={-1} rowData={{...}} classNames=\\"\\" columnId=\\"customfieldeval\\" cellData={0} columnData={{...}} customFieldComponents={{...}} updateData={[Function: updateData]} acceptValue={[Function: acceptValue]} resetValue={[Function: resetValue]} focused={false} editMode={false} isDisabled={false} />
    </DataTableCellHoverWrap>
  </div>
</div>"
`;
